### PR TITLE
feat: add config to use reasonable default expression instead of todo! when filling missing fields

### PR DESCRIPTION
**Pull request by bnjjj**
Use `Default::default()` in struct fields when we ask to fill it instead of putting `todo!()` for every fields

before:

```rust
pub enum Other {
    One,
    Two,
}

pub struct Test {
    text: String,
    num: usize,
    other: Other,
}

fn t_test() {
    let test = Test {<|>};
}
``` 

after: 

```rust
pub enum Other {
    One,
    Two,
}

pub struct Test {
    text: String,
    num: usize,
    other: Other,
}

fn t_test() {
    let test = Test {
        text: String::new(),
        num: 0,
        other: todo!(),
    };
}
```

### CODE REVIEW

**Comment by jhgg** [comment]
I would prefer this be an config personally, the todo!() is very handy at letting me know what I need to fill out, because not all the time is default the relevant value. 

I do wonder if we should have an assist that converts from `todo!()` -> `Default::default()` for situations where you *do* want the default.

**Comment by bnjjj** [comment]
The assist to convert `todo!()` -> `Default;;default()` would be very boring and inefficient IMHO. Maybe it could be configurable indeed. I often have this situation when I want to quickly fill a struct to test something or just because default values are good enough. Maybe @lnicola or @Veykril have ideas ?

**Comment by Veykril** [comment]
Making this configurable sounds like a good idea to me, the conversion assist could be done as well but shouldn't remove the need for a config here.

Note that there are few more "modes" that would come to mind here(not relevant to the PR aside from designing the config API space) like emitting `()` instead(this was the very first behaviour we had) or "sensible" defaults like `String::new()` or `0` etc.

**Comment by bnjjj** [comment]
Let's say we have a configuration named `fillStructFieldsMode` with these 2 modes for now `todo` and `defaultImpl` Do you have better ideas ? Or better names ? I'm bad at naming things

**Comment by Veykril** [comment]
`structFieldFill` maybe(I personally don't feel like `mode` adds any descriptive meaning to a config key)? Can't say I know a good name for this one here either(though it can always be changed later if someone comes up with a better one).

For the values `todo` and `defaultImpl` seem alright names to me.

I also just realized completions could make use of this config as well for snippet default, so this config has overlap with completions and diagnostics at least, so the quuestion arises under what config grouping we want to put this as well 🤔 Will put up a question for that on the zulip stream.

**Comment by Veykril** [comment]
Also linking this here as it may be relevant in the future(turns out assists also want this)

https://github.com/rust-analyzer/rust-analyzer/pull/10105#issuecomment-910114989
> This should be consitent with `fill_match_arms` first and foremost. That uses `todo!()` at the moment.
> 
> Either way, we should extract the code for "come up with default expression for type" into assist/utils, and use that in both places.
> 
> As to what's that expression is to be, I see the following choices:
> 
>     * `()` as the smallest thing
> 
>     * `todo!()` short, and semantically correct thing
> 
>     * default value for a type (`0`, `None`, `HashMap::default()`, depending on the type in question).
> 
> 
> I think `todo!()` is probably the best choice -- it's reasonably compact, reasonably correct, and robust (it doesnt' depend on type of the thing, which is an advantage, as IDE always behaves predictably).

**Comment by Veykril** [comment]
So ye since `add_missing_match_arms` could also make use of a default filler here I propose adding the config as something like `assist.exprFillDefault`(name pending) with `todo` and `defaultImpl`, that can be re-used by multiple things later on then(not required for this PR of course)

**Comment by bnjjj** [comment]
Ok great ! Thanks for your help and suggestions, I will implement the config support, and then in another PR later I'll add also the support for `add_missin_match_arms`

**Comment by bnjjj** [comment]
Should be good :)

**Comment by matklad** [comment]
I strongly prefer this to be `todo!()` without any configuration -- you almost *never* want to spell `Default::default()`, as it's usually better to use `HashMap::new` or `MyType::default`. 

I could see us adding a config for filligin "reasonable defaults", liken `0` for numbers, `None` for option, etc.

**Comment by Veykril** [comment]
This was discussed a bit with the same points that were raised here on zulip https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/.E2.9C.94.20Config.20Keys.20when.20applying.20to.20multiple.20groups

So consensus is that we should not offer merely `Default::default`, but instead some form of reasonable defaults.
Quoting matklad from the zulip channel:
> Yeah, that's my reasoning exactly. "reasonable" means Pareto-optimal -- we can't do anything strictly better. And there's a possible Pareto-improvement to Default::default -- instead of literal Default::default, use appropriate default value for each type ("zero" literals for primitive types, T::new() if that exists, T::default if that exists).

**Comment by bnjjj** [comment]
Ok will try to implement it this week ! Thanks for your suggestions

**Comment by bnjjj** [comment]
I implemented what you suggested. I still have a bug here https://github.com/rust-analyzer/rust-analyzer/pull/11145/files#diff-9636e6a0f0502bec621cdb4da24a9c1c531ae9479dfb859f85d5e09bba39a06aR161-R176 it never finds a `new` method. If you have ideas about what's wrong let me know.

**Comment by bnjjj** [comment]
I find my mistake, it should be ready to review/merge

File: crates/ide_diagnostics/src/handlers/missing_fields.rs [hunk 1]
```diff
@@ -102,6 +121,69 @@ fn fixes(ctx: &DiagnosticsContext<'_>, d: &hir::MissingFields) -> Option<Vec<Ass
...
+) -> Option<ast::Expr> {
+    if let Some(builtin_ty) = ty.as_builtin() {
+        if builtin_ty.is_int() || builtin_ty.is_uint() {
+            return Some(make::ext::zero_number());
+        }
+        if builtin_ty.is_float() {
+            return Some(make::ext::zero_float());
+        }
+        if builtin_ty.is_char() {
+            return Some(make::ext::empty_char());
```
Comment (Veykril) [inline crates/ide_diagnostics/src/handlers/missing_fields.rs:146]: Would probably make sense to use `\x00` as the default char here, as that is what the default impl of it returns as well

File: crates/ide_diagnostics/src/handlers/missing_fields.rs [hunk 2]
```diff
@@ -102,6 +121,69 @@ fn fixes(ctx: &DiagnosticsContext<'_>, d: &hir::MissingFields) -> Option<Vec<Ass
...
+    let krate = ctx.sema.to_module_def(d.file.original_file(ctx.sema.db))?.krate();
+    let module = krate.root_module(ctx.sema.db);
+
+    // Look for a ::new() associated function
+    let has_new_func = ty
+        .iterate_assoc_items(ctx.sema.db, krate, |assoc_item| {
+            if let AssocItem::Function(func) = assoc_item {
+                if func.name(ctx.sema.db) == known::new
+                    && func.assoc_fn_params(ctx.sema.db).is_empty()
+                    && func.self_param(ctx.sema.db).is_none()
```
Comment (Veykril) [inline crates/ide_diagnostics/src/handlers/missing_fields.rs:162]: This is superfluous, if `assoc_fn_params` is empty then there is no self param either

File: crates/rust-analyzer/src/config.rs [hunk 3]
```diff
@@ -1253,6 +1268,14 @@ fn field_props(field: &str, ty: &str, doc: &[&str], default: &str) -> serde_json
                 "Merge imports from the same module into a single `use` statement."
             ],
         },
+        "ExprFillDefaultDef" => set! {
+            "type": "string",
+            "enum": ["todo", "defaultImpl"],
+            "enumDescriptions": [
+                "Fill missing elements with 'todo' macro",
+                "Fill missing elements with T::default()"
```
Comment (Veykril) [inline crates/rust-analyzer/src/config.rs:1276]:
```suggestion
                "Fill missing expressions with reasonable defaults, `new` or `default` constructors."
```

**Review by Veykril — COMMENTED** [review]
Would probably be good to replace `defaultImpl` mentions with just `default` now, and where you are mentioning `elements` in the docs `expressions` would be more appropriate

File: crates/rust-analyzer/src/config.rs [hunk 4]
```diff
@@ -1253,6 +1268,14 @@ fn field_props(field: &str, ty: &str, doc: &[&str], default: &str) -> serde_json
                 "Merge imports from the same module into a single `use` statement."
             ],
         },
+        "ExprFillDefaultDef" => set! {
+            "type": "string",
+            "enum": ["todo", "defaultImpl"],
+            "enumDescriptions": [
+                "Fill missing elements with 'todo' macro",
```
Comment (lnicola) [inline crates/rust-analyzer/src/config.rs:1275]: Nit:

```suggestion
                "Fill missing expressions with the 'todo' macro",
```

File: crates/rust-analyzer/src/config.rs [hunk 5]
```diff
@@ -45,6 +45,8 @@ use crate::{
 // parsing the old name.
 config_data! {
     struct ConfigData {
+        /// How assists will fill missing elements in an expression.
```
Comment (bnjjj) [inline crates/rust-analyzer/src/config.rs:48]: @lnicola @Veykril better idea for this description ?

File: editors/code/package.json [hunk 6]
```diff
@@ -378,6 +378,19 @@
                     "markdownDescription": "Optional settings passed to the debug engine. Example: `{ \"lldb\": { \"terminal\":\"external\"} }`"
                 },
                 "$generated-start": {},
+                "rust-analyzer.assist.exprFillDefault": {
+                    "markdownDescription": "How assists will fill missing elements in an expression.",
```
Comment (lnicola) [inline editors/code/package.json:382]:
```suggestion
                    "markdownDescription": "Placeholder for missing expressions in assists.",
```

maybe? Not sure.

File: editors/code/package.json [hunk 7]
```diff
@@ -1494,4 +1507,4 @@
             ]
         }
     }
-}
+}
```
Comment (lnicola) [inline editors/code/package.json:1510]: Nit: let's keep the final newline.

File: editors/code/package.json [hunk 8]
```diff
@@ -378,6 +378,19 @@
...
+                "rust-analyzer.assist.exprFillDefault": {
+                    "markdownDescription": "How assists will fill missing elements in an expression.",
+                    "default": "todo",
+                    "type": "string",
+                    "enum": [
+                        "todo",
+                        "default"
+                    ],
+                    "enumDescriptions": [
+                        "Fill missing expressions with the 'todo' macro",
```
Comment (lnicola) [inline editors/code/package.json:390]:
```suggestion
                        "Fill missing expressions with the `todo` macro",
```

sorry, missed it the first time.

File: editors/code/package.json (hunk 7 above)
Comment (bnjjj) [reply editors/code/package.json:1510]: Yes sorry, it was automatically formatted by my code editor

File: editors/code/package.json (hunk 6 above)
Comment (bnjjj) [reply editors/code/package.json:382]: LGTM thank you

**Comment by bnjjj** [comment]
All done ! Thanks for your reviews ! It should be good :rocket:

**Review by Veykril — APPROVED** [review]
Thanks!
bors r+

**Pull request merged**